### PR TITLE
Added support for configuring postgres SSL support in the operator

### PR DESCRIPTION
**Pull request by jroper**
The work necessary in the proxy will be in a different PR.

### CODE REVIEW

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 1]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
...
+	// SslMode is the mode that should be used for SSL. Valid values are disable, allow, prefer, require,
+	// verify-ca and verify-full. Defaults to disable.
+	// +optional
+	Mode PostgresSSLMode `json:"mode,omitempty"`
+
+	// Secret is the secret that contain any SSL certificates or keys.
+	// +optional
+	Secret *corev1.LocalObjectReference `json:"secret,omitempty"`
+
+	// SSLRootCert is the key of the root certificate in the secret. This must be a PEM encoded X509 certificate.
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:100]:
```suggestion
	// RootCert is the key of the root certificate in the secret. This must be a PEM encoded X509 certificate.
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 2]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
 	DatabaseKey string `json:"databaseKey,omitempty"`
 }
 
+// PostgressSSL is the configuration for Postgres SSL connections.
+type PostgresSSL struct {
+	// SslMode is the mode that should be used for SSL. Valid values are disable, allow, prefer, require,
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:91]:
```suggestion
	// Mode is the mode that should be used for SSL. Valid values are disable, allow, prefer, require,
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go (hunk 1 above)
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:100]: Fields do not match comments: RootCert <=> SSLRootCert
Are those comments generated from the CRD YAML files?

Comment (jroper) [reply cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:100]: No, it's the other way around, the CRD YAML files use these comments.

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go (hunk 2 above)
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:91]:
```suggestion
	// Mode is the mode that should be used for SSL. Valid values are disable, allow, prefer, require,
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go (hunk 1 above)
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:100]:
```suggestion
	// RootCert is the key of the root certificate in the secret. This must be a PEM encoded X509 certificate.
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 3]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
...
+
+	// Secret is the secret that contain any SSL certificates or keys.
+	// +optional
+	Secret *corev1.LocalObjectReference `json:"secret,omitempty"`
+
+	// SSLRootCert is the key of the root certificate in the secret. This must be a PEM encoded X509 certificate.
+	// +optional
+	RootCert string `json:"rootCert,omitempty"`
+
+	// SSLCert is the key of the client certificate in the secret. This must be a PEM encoded X509 certificate.
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:104]:
```suggestion
	// Cert is the key of the client certificate in the secret. This must be a PEM encoded X509 certificate.
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 4]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
...
+	// It is ignored if using a PKCS-12 keychain.
+	// +optional
+	Cert string `json:"cert,omitempty"`
+
+	// SSLKey is the key of the private key for the client to use in the secret. This must be a PKCS-12 keychain or a
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:109]:
```suggestion
	// Key is the key of the private key for the client to use in the secret. This must be a PKCS-12 keychain or a
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 5]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
...
+	// PKCS-8 DER encoded key. If encrypted, the password should be specified in sslPassword.
+	// +optional
+	Key string `json:"key,omitempty"`
+
+	// SSLPassword is the password to use to decrypt the sslKey.
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:114]:
```suggestion
	// Password is the password to use to decrypt the sslKey.
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 6]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
...
+	// +optional
+	Password string `json:"password,omitempty"`
+}
+
+// PostgresSSLMode is the mode for SSL in postgres.
+type PostgresSSLMode string
+
+const (
+	// PostgresSSLModeDisable disable SSL, use if you don't care about security and don't want to pay the overhead
+	// for encryption
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:124]:
```suggestion
	// for encryption.
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 7]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
...
+	PostgresSSLModeDisable PostgresSSLMode = "disable"
+
+	// PostgresSSLModeAllow allow SSL, use if you don't care about security but will pay the overhead for encryption
+	// if the server insists on it
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:128]:
```suggestion
	// if the server insists on it.
```

File: cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go [hunk 8]
```diff
@@ -82,6 +86,65 @@ type PostgresCredentials struct {
...
+	PostgresSSLModeAllow PostgresSSLMode = "allow"
+
+	// PostgresSSLModePrefer prefer SSL, use if you don't care about encryption but will pay the overhead of encryption
+	// if the server supports it
```
Comment (marcellanz) [inline cloudstate-operator/pkg/apis/v1alpha1/statefulstore_types.go:132]:
```suggestion
	// if the server supports it.
```

**Review by pvlugter — APPROVED** [review]
I'm assuming this is good to go. I'll merge now.

**Pull request merged**